Support parsing block_states with version greater than 1.18

suzakuwcx opened the pull request:
Fix relate issue: #10, cite: https://minecraft.wiki/w/Chunk_format

But in contrast, this pr will cause that deprecating the support for old versions. So maybe creating a new branch is a better idea?

![image](https://github.com/MestreLion/mcworldlib/assets/73207484/98872c28-ad00-4231-823b-d08e5b2515b5)

suzakuwcx committed:
Support parsing block_states with version greater than 1.18

MestreLion commented:
Thank you, this is such a welcome contribution!

Btw, is the difference only in the capitalization of names (`sections` vs `Sections`) or were there other structural changes? If it is, it might be easy to create a unified `get_blocks()` that is compatible with both the old and the new standard. Even some minor structural changes (`palette` / `data` being under `block_state` or directly  in `Sections`) might be handled without too complex logic.

If not, I'd prefer the old function to remain intact, possibly provisionally renamed as `get_blocks_old()`, remove any compatibility logic in `get_blocks()`, and add documentation (as docstrings or plain comments) on which game versions each one operates.

In the future, if desired, `get_blocks()` may become a wrapper that automatically chooses the appropriate function based on the actual save file version.

MestreLion commented:
Also, would you mind adding a "Fix #10" suffix to the commit message so that bug is properly closed and linked to this fix?

suzakuwcx commented:
Thanks, I would try your suggestion to make backward compatible and update suffix.

About difference, it is not the only capitalization, base on my experiment on 1.21 and wiki description, the different is:
- Name change: "section" -> "Section", "BlockStates" ->  "block_states", "Palette" -> "palette"
- Structure change: "palette" now is under "block_state" (before is under "section")
- If only one block in "palette", the "data" field will be remove
-  sometime a strange "Y=-5" Section will appear and this is useless

suzakuwcx committed:
Support auto version detection and use correct method on parsing chunk

suzakuwcx committed:
Prevent from running the test script due to the need to switch directories and installation.

suzakuwcx committed:
Fix: 'chunk.get_blocks' function using 'yield' instead of 'return'

suzakuwcx committed:
Fix chunk.py: '_decode_blockstates' will failed if palette number greater than 16

suzakuwcx committed:
Revert commit a373585c9bbd69a8646d1af7212485ec5398b16d

suzakuwcx committed:
Fix #10

- Upgrade supportting parsing version to 1.21
- Fix '_decode_blockstates' if palette number greater than 16

suzakuwcx commented:
I test the test case and now it goes well

blocks()
![image](https://github.com/user-attachments/assets/29130ef4-627c-4d64-beca-7bdd6a29cd1c)

pretty()
![image](https://github.com/user-attachments/assets/ec22ebf6-9e07-4fb5-8d09-d13fe787680b)

SorkoPiko commented:
When is this getting merged? I would love to see this implemented, as there are currently no working MC world manipulation python libraries.

suzakuwcx committed:
Fix #10

Because that the index of palette array begins from 0, so the maximux
index is 1 less than palette length

MestreLion commented:
@SorkoPiko 
> When is this getting merged? I would love to see this implemented, as there are currently no working MC world manipulation python libraries.

I didn't have the time to properly review or test it yet. The new block state format changes more than just renaming/reorganizing keys in the NBT structure, and this PR also deals with backward-compatibility at run time, so not very trivial. If anyone can/wants to test this, or even step up as a co-maintainer of this project, I would appreciate!

corgi-in-tights commented:
To anyone in the future: PR seems to work for 1.21.8. (atleast the few methods I tested), recommend just cloning and replacing your current mcworldlib folder with the one from suzaku's repo

Starrah commented on mcworldlib/chunk.py:
```diff
@@ -44,8 +44,36 @@ def entities(self):
-    def get_blocks(self):
-        """Yield a (Y, Palette, BlockState Indexes Array) tuple for every chunk section.
+
+    def is_version_1_21(self):
```
```suggestion
    def is_version_1_18(self):
```
In fact (as described in the Wiki and I also tested by myself), the change of name "Sections" to "sections" and other name changes is applied from 1.18 rather than 1.21.

So the name can be changed?

Starrah commented:
I tested this PR on:

- [x] 1.17.1 (The last version of the old format)
- [x] 1.18.2 (The changes of the chunk format related to this PR are since 1.18)
- [x] 1.20.1
- [x] 1.21.1

All of them work well.

So I think it is OK to merge this PR. The function work good, though it has some code names that may confuse others as I commented above.

Starrah commented:
> If anyone can/wants to test this, or even step up as a co-maintainer of this project

Besides, I'm also interested in step up as a co-maintainer of this project, (as it is very useful and helped me find my [Little Maid](https://www.mcmod.cn/item/186726.html)'s tomb position when I'm offline, on 1.20.1.) @MestreLion
